doc: Add anyone-can-pay lock specification (#205)
nervosnetwork/rfcs: rfcs/0026-anyone-can-pay/0026-anyone-can-pay.md

---
Number: "0026"
Category: Standards Track
Status: Proposal
Author: Xuejie Xiao
Organization: Nervos Foundation
Created: 2020-09-03
---

# Anyone-Can-Pay Lock

> This RFC describes a new lock script for CKB that can accept any amount of payment. Previously, one can only transfer to another user at least 61 CKBytes when using the default lock, possibly more when using other lock scripts or type scripts. This is becoming a bigger problem when UDT support lands in CKB: a naive UDT transfer operation will not only require UDTs, but CKByte to keep the UDTs in a cell as well.

Reviewer: > any amount of payment
any amount of CKB or UDT payment

Here we try to solve the problem by introducing a new anyone-can-pay lock script, which can be unlocked not only by the validation of a signature, but also by accepting any amount of payment. This way, a user should be able to send any amount of CKBytes or UDTs to a cell using anyone-can-pay lock instead of always creating a new cell. It thus provides a solution to both problems above.

## Script Structure

The anyone-can-pay lock is built upon the default secp256k1-blake2b-sighash-all lock with additions to the script args part. The new anyone-can-pay lock can accept any of the following script args format:

```
<20 byte blake160 public key hash>
<20 byte blake160 public key hash> + <1 byte CKByte minimum>

> <20 byte blake160 public key hash> + <1 byte CKByte minimum> + <1 byte UDT minimum>

Reviewer: There’s no `+` in the following examples, maybe they also can be removed here.

The RFC does not talk about how the two minimums are encoded.

Reviewer: The minimums may lead to integer overflow, The RFC has to specify the maximum allowed values. From the implementation one uses u64 and another uses u128

```
  uint64_t min_ckb_amount;
  uint128_t min_udt_amount;
```

Reviewer: Added a new comment on this line https://github.com/nervosnetwork/rfcs/pull/205/files/722dba240f8c33338a67a844000c691fcb51a979#r483358842

Author: Personally I feel like it could be up to the contracts to handle integer overflows. There might be an implementation that choose to implement arbitrary long integer arithmetic, which won't have overflowing problems.

Reviewer: This explanation can be added to the RFC explicitly, however, I don’t have the strong opinion to block the PR because of this.

```

> The additions of CKByte & UDT minimums enforces the minimal amount one can transfer to the anyone-can-pay lock. This provides a mitigation against DDoSing on cell level: if a cell is setup using anyone-can-pay lock, an attacker can keep creating transactions that transfer only 1 shannon or 1 UDT to the cell, make it difficult for the cell owner to claim the tokens stored in the cell. By providing a minimal transfer amount, a user can raise the attacking cost, hence protecting his/her own cells against DDoS attacks. This mechanism won't prevent all kinds of DDoS of course, but it serves as a quick solution to mitigate cheaper ones.

Reviewer: ```suggestion
The additions of CKByte & UDT minimums enforce the minimal amount that one can transfer to the anyone-can-pay lock. This provides a mitigation against DDoSing on cell level: if a cell is setup using the anyone-can-pay lock, an attacker can keep creating transactions that transfer only 1 shannon or 1 UDT to the cell, making it difficult for the cell owner to claim the tokens stored in the cell. By providing a minimal transfer amount, a user can raise the attacking cost, hence protecting his/her own cells against DDoS attacks. This mechanism won't prevent all kinds of DDoS of course, but it serves as a quick solution to mitigate cheaper ones.
```

* If 4 is stored in UDT base unit minimum, it means the minimal amount that can be accepted by the cell is 10000 UDT base units.

Note the minimum fields are completely optional. If a minimum is not provided, we will treat the minimum value as 0, meaning no minimum is enforced on the transfer operation.

## UDT Interpretation

The anyone-can-pay lock assumes that the locked cell follows the [Simple UDT specification](https://talk.nervos.org/t/rfc-simple-udt-draft-spec/4333), thus the cell 1) has a type script; 2) has at least 16 bytes in the cell data part. Its up to the user to ensure one only uses anyone-can-pay lock with a type script implementing Simple UDT specification.

## Unlock Rules

The anyone-can-pay lock will work following the rules below:

1. If a signature is provided, it works exactly as the default secp256k1-blake2b-sighash-all lock, if a signature is provide in witness and can be validated, the lock returns with a success state.

    1.a. If the provided signature fails validation, the lock returns with an error state

2. If a signature is not provided, the lock continues with the added anyone-can-pay logic below:

>     2.a. It loops through all input cells using the current anyone-can-pay lock script(notice here the lock script we refer to include public key hash, meaning if a transaction contains 2 cells using the same anyone-can-pay lock code, but different public key hash, they will be treated as different lock script, and each will perform the script unlock rule checking independently), if 2 input cells are using the same type script, or are both missing type scripts, the lock returns with an error state

Reviewer: Should we mention that different minimums also lead to different lock script?

Author: Added in the above structure section

### Create an Anyone-can-pay Cell

```
Inputs:
    Normal Cell:
        Capacity: 1000 CKBytes
        Lock:
            code_hash: secp256k1_blake2b lock
            args: <public key hash A>
Outputs:
    Anyone-can-pay Cell:
        Capacity: 999.99 CKBytes
        Type:
            code_hash: simple udt lock
            args: <owner lock C>
        Lock:
            code_hash: anyone-can-pay lock
            args: <public key hash B> <CKByte minimum: 9> <UDT minimum: 5>
        Data:
            Amount: 0 UDT
Witnesses:
    <valid signature for public key hash A>
```

Note here we assume 0.01 CKByte is paid as the transaction fee, in production one should calculate the fee based on factors including transaction size, running cycles as well as network status. 0.01 CKByte will be used in all examples as fees for simplicity. The new anyone-can-pay cell created by this transaction impose a minimum transfer value of 10^9 shannons (10 CKBytes) and 10^5 UDT base units respectively.

### Unlock via Signature

```
Inputs:
    Anyone-can-pay Cell:
        Capacity: 1000 CKBytes
        Lock:
            code_hash: anyone-can-pay lock

>             args: <public key hash A> <CKB minimum: 2>

Reviewer: Use consistent name

```suggestion
            args: <public key hash A> <CKByte minimum: 2>
```

Outputs:
    Normal Cell:
        Capacity: 999.99 CKBytes
        Lock:
            code_hash: secp256k1_blake2b lock
            args: <public key hash B>
Witnesses:
    <valid signature for public key hash A>
```

When a signature is provided, the cell can be unlocked in anyway the owner wants, anyone-can-pay lock here just behaves as a normal cell. In this example an anyone-can-pay cell is converted back to a normal cell.

### Unlock via CKB Payment on Cells with No Type Script

```
Inputs:
    Deposit Normal Cell:
        Capacity: 500 CKBytes
        Lock:
            code_hash: secp256k1_blake2b lock
            args: <public key hash B>
    Anyone-can-pay Cell:
        Capacity: 1000 CKBytes
        Lock:
            code_hash: anyone-can-pay lock

>             args: <public key hash A> <CKBytes minimum: 2>

Reviewer: ```suggestion
            args: <public key hash A> <CKByte minimum: 2>
```

Please do a search and replace

```

Here we are depositing 1 UDT to the anyone-can-pay cell. Because theres no extra arguments in the anyone-can-pay lock script except a public key hash, the cell enforces no minimum on the CKByte or UDT one can transfer, a transfer of 1 UDT will be accepted here.

### Unlock via CKByte Payment With Minimums

```
Inputs:
    Deposit Normal Cell:
        Capacity: 500 CKBytes
        Lock:
            code_hash: secp256k1_blake2b lock
            args: <public key hash B>
        Type:
            code_hash: simple udt lock
            args: <owner lock C>
        Data:
            Amount: 200000 UDT
    Anyone-can-pay Cell:
        Capacity: 1000 CKBytes
        Lock:
            code_hash: anyone-can-pay lock
            args: <public key hash A>
        Type:
            code_hash: simple udt lock
            args: <owner lock C> <CKBytes minimum: 9> <UDT minimum: 5>
        Data:
            Amount: 3000 UDT
Outputs:
    Deposit Change Cell:
        Capacity: 489.99 CKBytes
        Lock:
            code_hash: secp256k1_blake2b lock
            args: <public key hash B>
        Type:
            code_hash: simple udt lock
            args: <owner lock C>
        Data:

>             Amount: 199999 UDT

Reviewer 2: ```suggestion
            Amount: 200000 UDT
```
